### PR TITLE
feat: added indexed priority queues

### PR DESCRIPTION
**Pull request by ron0studios**
also added tests for the implementation. See the file for more details

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

Im not sure whether I should rename the filename as something like `queue-priority-index` instead of `indexed-priority-queue` for the sake of file grouping.

### CODE REVIEW

File: data_structures/indexed_priority_queue.cpp [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+ */
+
+
+#include <vector> /// std::vector
+#include <utility> /// std::pair
+#include <unordered_map> /// std::std::unordered_map
+#include <cmath> /// std::floor
+
+#include <string> /// tests
+#include <cassert>   /// for assert
```
Comment (Panquesito7) [inline data_structures/indexed_priority_queue.cpp:28]:
```suggestion
#include <vector> /// for std::vector
#include <utility> /// for std::pair
#include <unordered_map> /// for std::unordered_map
#include <cmath> /// for std::floor
#include <string> /// for std::string
#include <cassert>   /// for assert
```

File: data_structures/indexed_priority_queue.cpp [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+			while(i >= 0 && val[im[node]] < val[im[i]])
+			{
+				swap(i,node);
+				node = i;
+				i = std::floor((node-1)/2);
+			}
+		}
+};
+
+static void test() {
```
Comment (Panquesito7) [inline data_structures/indexed_priority_queue.cpp:200]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
static void test() {
```

File: data_structures/indexed_priority_queue.cpp [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+		{"python",3},
+		{"Ron0Studios",10}
+	};
+
+	assert(heapB == expectedB);
+}
+
+int main()
+{
+	test();
```
Comment (Panquesito7) [inline data_structures/indexed_priority_queue.cpp:253]:
```suggestion
    test();  // run self-test implementations
```

File: data_structures/indexed_priority_queue.cpp [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+		{"F#",0},
+		{"",0},
+		{"python",3},
+		{"Ron0Studios",10}
+	};
+
+	assert(heapB == expectedB);
+}
+
+int main()
```
Comment (Panquesito7) [inline data_structures/indexed_priority_queue.cpp:251]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main()
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Comment by ron0studios** [comment]
Just made the relevant changes, thanks!

**Comment by ron0studios** [comment]
Is there anything else I need to add?

**Comment by ron0studios** [comment]
Sorry, @Panquesito7 did I miss anything out? For some reason it still says that there are changes requested even though they've been resolved? Thanks,

**Comment by ron0studios** [comment]
@Panquesito7 Could you do another review please? It's been a couple of days and I've made the relevant changes. If there is anything that needs adding, please do let me know! Thank you

**Comment by ron0studios** [comment]
please @Panquesito7 could you review this, or anyone else, if possible? Sorry if I'm coming off as impatient, I would just like some forewarning as to how long the process will atleast take. Thank you!

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix the [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/actions/runs/3274169690/jobs/5388637894) warnings. Let us know if you need any help. 🙂

File: data_structures/indexed_priority_queue.cpp [hunk 5]
```diff
@@ -0,0 +1,258 @@
+/**
+ * @file
+ * @brief This is an implementation of an min indexed priority queue,
```
Comment (Panquesito7) [inline data_structures/indexed_priority_queue.cpp:3]: Provide a Wikipedia link in Markdown format (if available/possible).
If there's no Wikipedia link, add another web source for algorithm explanation. 🙂

File: data_structures/indexed_priority_queue.cpp [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+			while(i >= 0 && val[im[node]] < val[im[i]])
+			{
+				swap(i,node);
+				node = i;
+				i = std::floor((node-1)/2);
+			}
+		}
+};
+
+static void test() {
```
Comment (Panquesito7) [inline data_structures/indexed_priority_queue.cpp:199]: Still not solved and should be fixed.
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
static void test() {
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix the [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/actions/runs/3278016892/jobs/5395939222) warnings. Let us know if you need any help with that.

File: data_structures/indexed_priority_queue.cpp (hunk 5 above)
Comment (ron0studios) [reply data_structures/indexed_priority_queue.cpp:3]: just added a link to a Wikipedia page!

File: data_structures/indexed_priority_queue.cpp (hunk 6 above)
Comment (ron0studios) [reply data_structures/indexed_priority_queue.cpp:217]: Apologies, added a description!

**Comment by ron0studios** [comment]
@Panquesito7 the clang tidy errors should also be resolved i think!

**Comment by ron0studios** [comment]
the changes have been resolved i think. is there anything else i need to add?

File: data_structures/indexed_priority_queue.cpp [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+ * @author [ron0studios](https://github.com/ron0studios)
+ */
+
+
+#include <vector>          /// for std::vector
+#include <utility>        /// for std::pair
+#include <unordered_map> /// for std::unordered_map
+#include <cmath>        /// for std::floor
+#include <string>       /// for std::string
+#include <cassert>      /// for assert
```
Comment (Panquesito7) [inline data_structures/indexed_priority_queue.cpp:28]:
```suggestion
#include <cassert>      /// for assert
#include <iostream>    /// for IO operations
```

File: data_structures/indexed_priority_queue.cpp [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+	}
+
+	std::vector<std::pair<std::string,int>> expectedB = {
+		{"F#",0},
+		{"",0},
+		{"python",3},
+		{"Ron0Studios",10}
+	};
+
+	assert(heapB == expectedB);
```
Comment (Panquesito7) [inline data_structures/indexed_priority_queue.cpp:265]:
```suggestion
	assert(heapB == expectedB);
	std::cout << "All tests have successfully passed!\n";
```

**Review by Panquesito7 — APPROVED** [review]
Awesome work! Thank you. 🚀

File: data_structures/indexed_priority_queue.cpp [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+                         std::vector<std::pair<T1, T2>>()) {
+            for(auto const &i : keyval){
+				insert(i.first, i.second);
+			}
+        }
+
+        /**
+		 * @brief check if the queue contains a certain key
+		 * @param key the key to check for
+	     */
```
Comment (tjgurwara99) [inline data_structures/indexed_priority_queue.cpp:65]: The commenting style and spacing looks weird to me. Could you please correctly align stuff and follow the style guide?

Comment (ron0studios) [reply data_structures/indexed_priority_queue.cpp]: Hi [tjgurwara99](https://github.com/tjgurwara99)! I just made the relevant changes. It looked perfectly fine in my editor, but showed up odd on github, must be a tabs/spacing issue on my end :)

**Review by tjgurwara99 — APPROVED** [review]

File: data_structures/indexed_priority_queue.cpp [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+ * the basis of a hashmap, providing its benefit of having O(1) searching
+ * and log(N) modification/removal of ANY key in the queue (not only the front)
+ * but also the benefits of a priority queue for polling min values in 
+ * logarithmic time or having its heap properties. The keys in this 
+ * implementation are the unique identifiers or stored values of the 
+ * queue, whilst the values refer to the comparable that will be used
+ * to prioritise different items. 
+ * More information can be found [here](https://en.wikipedia.org/wiki/Priority_queue)
+ * refer to william fiset's video for a video explanation:
+ * https://youtu.be/jND_WJ8r7FE
```
Comment (Panquesito7) [inline data_structures/indexed_priority_queue.cpp:18]:
```suggestion
 * Refer to [William Fiset's video](https://youtu.be/jND_WJ8r7FE) for a good explanation.
```

File: data_structures/indexed_priority_queue.cpp [hunk 11]
```diff
@@ -0,0 +1,278 @@
+/**
+ * @file
+ * @brief This is an implementation of an min indexed priority queue,
```
Comment (Panquesito7) [inline data_structures/indexed_priority_queue.cpp:3]:
```suggestion
 * @brief [Priority Queue](https://en.wikipedia.org/wiki/Priority_queue) implementation,
```

File: data_structures/indexed_priority_queue.cpp [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+		{30,0},
+		{290,1},
+		{5,2},
+		{29,10}
+	};
+
+	assert(heapA == expectedA);
+
+	// test round B (insertion, deletion and template params)
+	IPQ<std::string,int> ipqB;
```
Comment (Panquesito7) [inline data_structures/indexed_priority_queue.cpp:241]:
```suggestion
	data_structures::priority_queue::IPQ<std::string,int> ipqB;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Comment by ron0studios** [comment]
Hi can I resolve and commit the suggested changes to complete the pull request?

**Pull request closed without merging**